api/graphql/schema.graphql

```diff
@@ -727,6 +727,38 @@ type ExperienceSkillRecord {
   details: String
 }
 
+type JobPosterTemplate {
+  id: UUID!
+  referenceId: String! @rename(attribute: "reference_id")
+  name: LocalizedString!
+  description: LocalizedString!
+  supervisoryStatus: SupervisoryStatus! @rename(attribute: "supervisory_status")
+  stream: LocalizedPoolStream!
+  workDescription: LocalizedString @rename(attribute: "work_description")
+  tasks: LocalizedString!
+  keywords: LocalizedString!
```

Comment:
I can do that but if they are non-nullable in the database is there a reason we should remove that? They should be guaranteed to be there saves us from doing extra null checking on the frontend :thinking:

Reply:
GraphQL recommends trying to avoid non-nullable fields, mostly for reasons of maintaining backwards compatibility, and we've always followed that except for the most essential fields. (If we leave this nullable in graphql, its easy to change the nullability in the database, but its hard to change the implementation if we make it non-nullable here.)

Reply:
Okay, I agree wit that recommendation. I brought it up a while ago and was told we do want to use non-nullable fields where possible (i.e match the database). So are we going to stop doing that in the future I assume?

[make fields nullable in schema](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11468/commits/bec52ae41222ed4cf4c6338225dadd24074d844a)

Reply:
If this feature ends up working like the pool workflow, being nullable will be essential as some fields may not get pencilled in at first

Reply:
> If this feature ends up working like the pool workflow, being nullable will be essential as some fields may not get pencilled in at first

Yeah, I was thinking that exactly.

> I brought it up a while ago and was told we do want to use non-nullable fields where possible (i.e match the database). So are we going to stop doing that in the future I assume?

I thought the norm (for us) was always to use `!` very minimally in the schema, regardless of the database. Do you remember what we were talking about where we decided otherwise? If I've been inconsistent on this, I apologize.

Reply:
I was complaining about how codegen used to handle them and suggested we should use non-nullable fields minimally (i.e for IDs and little else) and rely more on the DB/validator to ensure non-null.

We never got a final decision but I was told to use non-nullable fields in the schema, for now, to avoid using validators/DB controls whenever we expect it to exist (so non-nullable/defaults)

Reply:
Okay, I do remember talking about that. My position was more that I didn't want our decisions about nullable/non-nullable fields in the schema to be dictated by typescript. I especially didn't want to modify our _existing_ schema to accommodate typescript, especially when I thought moving to fragments was a better solution.

Also, its worth mentioning that GraphQL suggests avoiding non-nullable fields on _types_, but goes the opposite for schema _inputs_. A non-nullable field on an _input_ type is great, and should be used wherever it makes sense. That's the one that helps us avoid validators. 

On types, you can go from nullable (default) to non-nullable, but not vice versa (at least without making a breaking change). For input types, its the opposite: going from a nullable to non-nullable field is a breaking change.

Reply:
Alright, yeah that makes sense. I think there was miscommunication at that point because I was referring to things that seem odd to be non-nullable. Like how we have the department number and name as non-nullable as well as alot of (most) relationships.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/3ac49ad35e2a8f96cbc9e21caa132b92592e85e9/api/graphql/schema.graphql#L468-L469

https://github.com/GCTC-NTGC/gc-digital-talent/blob/3ac49ad35e2a8f96cbc9e21caa132b92592e85e9/api/graphql/schema.graphql#L655

Are we still aiming to have relationships as non-nullable? A lot must exist so its not necessarily a misuse but we are not consistent with it either :thinking: